src/server/pingserver-rs/src/worker.rs

```diff
@@ -73,74 +74,81 @@ impl Worker {
             for event in events.iter() {
                 let token = event.token();
 
-                if token != self.waker_token {
-                    // event for existing session
-                    trace!("got event for session: {}", token.0);
+                // event for existing session
+                trace!("got event for session: {}", token.0);
 
-                    if event.is_error() {
-                        self.increment_count(&Stat::WorkerEventError);
-                        self.handle_error(token);
-                    }
+                if event.is_error() {
+                    self.increment_count(&Stat::WorkerEventError);
+                    self.handle_error(token);
+                }
 
-                    if event.is_readable() {
-                        self.increment_count(&Stat::WorkerEventRead);
-                        let _ = self.do_read(token);
-                    }
+                if event.is_readable() {
+                    self.increment_count(&Stat::WorkerEventRead);
+                    let _ = self.do_read(token);
+                }
 
-                    if event.is_writable() {
-                        self.increment_count(&Stat::WorkerEventWrite);
-                        self.do_write(token);
-                    }
+                if event.is_writable() {
```

Comment:
now that I think about it, maybe we should handle write event before read event if they are both present. The existing C logic has the same order but that's not optimal. Since write event only occurs when the last write was blocked, this allows us to flush the write buffer before handling (potential) new reads, and therefore reduces the chance of having to expand buffer due to new response on top of residual data.

Reply:
Reordered for admin/server/worker for consistency. Comments added to indicate reasoning for the order.